components/formats-gpl/src/loci/formats/in/KLBReader.java

```diff
@@ -0,0 +1,577 @@
+      dims_blockSize[i] = readUInt32();
+    }
+    blocksPerPlane = (int) (Math.ceil((float)getSizeX()/dims_blockSize[0]) * Math.ceil((float)getSizeY()/dims_blockSize[1]));
+    
+    numBlocks = 1;
+    for (int i=0; i < KLB_DATA_DIMS; i++) {
+      numBlocks *= Math.ceil((float)(dims_xyzct[i]) / (float)(dims_blockSize[i]));
+    }
+
+    headerSize = (long) ((KLB_DATA_DIMS * 12) + 2 + (numBlocks * 8) + KLB_METADATA_SIZE + 1);
+    
+    String order = core.get(getSeries()).dimensionOrder;
+    int[] ztc = FormatTools.getZCTCoords(order, getSizeZ(), getSizeC(), getSizeT(), getImageCount(), no);
+
+    // Calculate the first required block
+    //TODO: Correct calculation for T and C blocks
+    //int requiredBlockNum = (ztc[0] / dims_blockSize[2]) * (ztc[2] / dims_blockSize[3]) * (ztc[1] / dims_blockSize[4]);
+    int requiredBlockNum = (ztc[0] / dims_blockSize[2]);
+
+    // Mark the current file pointer to return after reading offsets
+    long filePoointer = in.getFilePointer();
+
+    // Seek to start of offsets and read required offsets
+    blockOffsets = new long[blocksPerPlane];
+    in.seek(offsetFilePointer + (requiredBlockNum * blocksPerPlane * 8));
+    for (int i=0; i < blocksPerPlane; i++) {
+      blockOffsets[i] = readUInt64();
+    }
+    in.seek(filePoointer);
+  }
+
+  // Helper methods
+
+  private void convertPixelType(CoreMetadata ms0, int pixelType) throws FormatException {
+    switch (pixelType) {
+      case UINT8_TYPE:
+        ms0.pixelType = FormatTools.UINT8;
+        break;
+      case UINT16_TYPE:
+        ms0.pixelType = FormatTools.UINT16;
+        break;
+      case UINT32_TYPE:
+        ms0.pixelType = FormatTools.UINT32;
+        break;
+      case UINT64_TYPE:
+      case INT64_TYPE:
+        ms0.pixelType = FormatTools.DOUBLE;
+        break;
+      case INT8_TYPE:
+        ms0.pixelType = FormatTools.INT8;
+        break;
+      case INT16_TYPE:
+        ms0.pixelType = FormatTools.INT16;
+        break;
+      case INT32_TYPE:
+        ms0.pixelType = FormatTools.INT32;
+        break;
+      case FLOAT32_TYPE:
+      case FLOAT64_TYPE:
+        ms0.pixelType = FormatTools.FLOAT;
+        break;
+      default:
+        throw new FormatException("Unknown pixel type: " + pixelType);
+    }
+    ms0.interleaved = ms0.rgb;
+  }
+
+  private int readUInt32() throws IOException {
+    byte[] b = new byte[4];
+    in.read(b, 0, 4);
+    ByteBuffer bb = ByteBuffer.wrap(b).order(ByteOrder.LITTLE_ENDIAN);
+    return bb.getInt();
+  }
+
+  private long readUInt64() throws IOException {
+    byte[] b = new byte[8];
+    in.read(b, 0, 8);
+    ByteBuffer bb = ByteBuffer.wrap(b).order(ByteOrder.LITTLE_ENDIAN);
+    return bb.getLong();
+  }
+
+  private PositiveFloat readFloat32() throws IOException {
+    byte[] b = new byte[4];
+    in.read(b, 0, 4);
+    ByteBuffer bb = ByteBuffer.wrap(b).order(ByteOrder.LITTLE_ENDIAN);
+    return new PositiveFloat((double) bb.getFloat());
+  }
```

Comment:
Is there a reason not to use the ```readInt```, ```readLong```, and ```readFloat``` methods in ```RandomAccessInputStream``` here?

Reply:
Yeah, this is because they are using the opposite endianness, I had originally tried to use RandomAccessInputStream for it

Reply:
Using https://github.com/ome/ome-common-java/blob/master/src/main/java/loci/common/RandomAccessInputStream.java#L256 as needed should handle that case; it's safe to swap endianness multiple times on an open stream.  See for example https://github.com/openmicroscopy/bioformats/blob/develop/components/formats-gpl/src/loci/formats/in/GatanReader.java#L683